### PR TITLE
react task ready for review

### PR DESCRIPTION
**Pull request by linkqwd**
# React, rick and morty api

Demo: [link](https://linkqwd.github.io/react-rick-and-morty/)
Code-base: [link](https://github.com/linkqwd/react-rick-and-morty/tree/master/src)

### CODE REVIEW

**Comment by AMashoshyna** [comment]
Great job! I have put together some comments, but I suggest that you commit your code to this repo too, so i can add comments to particular lines. But here is the first part:

1. `DataRetriever` and `ArraySerializer` are unfortunate names for functions: only class names and constructor function names should start with a capital letter. Also ArraySerializer does not actually work with array, but with HTML FormElement, which is confusing.
2. SerializeArray.js
- please avoid `var`s
- using Array.prototype.map or Array.prototype.forEach is preferable to for-loop, since it's more declarative.
- this method of error handling is unsafe:
```javascript
  async function query(address) {
    const response = await fetch(address);
    const responseData = await response.json();

    if (responseData.error) {
      throw new Error(responseData);
    } else {
      return responseData;
    }
  }
```
If the request fails (let's say, status 404 returns), the function will throw an error earlier than you expect, namely at the attempt to parse response body as JSON. Most often you want to check for errors by checking if the `Response.ok` is `true`.
- using strings as identifiers is error-prone because of typing mistakes, which IDE cannot help to avoid.
```javascript
 if (type === "byId") {
   //...
}
```
A good way to avoid is was shown by Alexander during WebSocket tutorial. You put the strings in an object and use properties of an object: 
```javascript
const REQUEST_TYPES = {
   BY_ID: 'byId',
   CUSTOM: 'custom'
}
// further in your code:
 if (type === REQUEST_TYPES.BY_ID) {
   //...
}

```
Same thing is applicable to statuses in `App` component state.
3. App.js

- method `ascDescArraySorter` - I suggest that this method return received array as default case, becasue otherwise the app will crash when child component will try to call `map` of `undefined`.

- `Array.prototype.find` maybe?
```javascript
    const characterIdSearchValue = formDataArray.filter(
      item => item.name === "name"
    )[0].value;
```

4. Aside.js
This component has a long `render()` method which contains repeating elements. Maybe you could refactor it?

5. Pagination.js
I suggest to shorten this condition:
```javascript
 if (props.pagesCount === 0 || props.pagesCount === undefined) return "";
```
All of the values mentioned are falsy.

**Comment by AMashoshyna** [comment]
One more important thing I forgot to mention: please avoid mutating function agruments as in [here](https://github.com/linkqwd/react-rick-and-morty/blob/master/src/components/App.js#L88). Create copies, if you need to/

**Comment by linkqwd** [comment]
Thank you a lot for such great review

File: submissions/linkqwd/react-rick-and-morty/utils/parseForm.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export default form => {
+  const additionalInputsToParse = {
+    searchInput: "search"
+  };
+
+  return Object.entries(form)
+    .filter(item => (item[1].name ? item : null))
+    .map(item => {
```
Comment (AMashoshyna) [inline submissions/linkqwd/react-rick-and-morty/utils/parseForm.js]: This function is very hard to read. Maybe put some logic aside in helpers?

File: submissions/linkqwd/react-rick-and-morty/utils/parseForm.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      return item[1].checked ||
+        item[1].type === additionalInputsToParse.searchInput
+        ? { name: item[1].name, value: item[1].value }
+        : null;
+    })
+    .filter(item => (item ? item : null));
```
[the remark this answers is not in the document]
Comment (mgdskr) [reply submissions/linkqwd/react-rick-and-morty/utils/parseForm.js]: a little bit shorter and more explicit variant`.filter(Boolean)`

**Comment by linkqwd** [comment]
Thank you. Very strong and helpful points.

**Pull request closed without merging**